polynote-runtime/src/main/scala/polynote/runtime/ValueRepr.scala

```diff
@@ -0,0 +1,249 @@
+package polynote.runtime
+
+import java.nio.ByteBuffer
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicInteger
+
+import sun.misc.Cleaner
+
+/**
+  * A plain text representation of a value
+  */
+final case class StringRepr(string: String) extends ValueRepr
+
+/**
+  * A MIME representation of a value. Binary values must be base64-encoded.
+  */
+final case class MIMERepr(mimeType: String, content: String) extends ValueRepr
+
+/**
+  * A binary representation of a value, encoded in Polynote's format (TODO)
+  */
+final case class DataRepr(dataType: DataType, data: ByteBuffer) extends ValueRepr
+
+/**
+  * An identifier for a "lazy" value. It won't be evaluated until the client asks for it by its handle.
+  * We use Ints for the handle, because JS needs to store the handle, and it can't store Long. We just can't have
+  * more than 2^32^ different handles... I think there would be bigger issues at that point.
+  */
+final case class LazyDataRepr private[polynote](handle: Int, dataType: DataType) extends ValueRepr {
+  def isEvaluated: Option[Boolean] = LazyDataRepr.getHandle(handle).map(_.isEvaluated)
+
+  def release(): Unit = {
+    LazyDataRepr.handles.remove(handle)
+  }
+}
+
+object LazyDataRepr {
+
+  private[polynote] class Handle(val handle: Int, val dataType: DataType, lazyData: => ByteBuffer) {
+    @volatile private var computedFlag: Boolean = false
+
+    lazy val data: ByteBuffer = {
+      computedFlag = true
+      lazyData
+    }
+
+    def isEvaluated: Boolean = computedFlag
```

Comment:
What is `isEvaluated` used (or planned to be used) for?

Reply:
Not sure... I guess so you could avoid forcing something but try to look at it if it were evaluated? I think the `LazyDataRepr` might go away; I thought there would be good uses for it but I'm not so sure anymore.